components/cldr-json-data-provider/src/support.rs

```diff
@@ -0,0 +1,69 @@
+use crate::CldrPaths;
+use icu_data_provider::iter::DataEntryCollection;
+use icu_data_provider::prelude::*;
+use std::convert::TryFrom;
+use std::sync::RwLock;
+
+pub(crate) trait DataKeySupport {
+    fn supports_key(data_key: &DataKey) -> Result<(), DataError>;
+}
+
+pub(crate) struct LazyCldrProvider<T> {
+    data_provider: RwLock<Option<T>>,
+}
+
+impl<'b, 'd, T> LazyCldrProvider<T>
+where
+    T: DataProvider<'d> + DataKeySupport + DataEntryCollection + TryFrom<&'b CldrPaths>,
+    <T as TryFrom<&'b CldrPaths>>::Error: 'static + std::error::Error,
+{
+    pub fn new() -> Self {
+        Self {
+            data_provider: RwLock::new(None),
+        }
+    }
+
+    pub fn try_load(
+        &self,
+        req: &DataRequest,
+        cldr_paths: &'b CldrPaths,
+    ) -> Result<Option<DataResponse<'d>>, DataError> {
+        if T::supports_key(&req.data_key).is_err() {
+            return Ok(None);
+        }
+        if self.data_provider.read().unwrap().is_none() {
+            self.data_provider.write().unwrap().replace(
+                T::try_from(cldr_paths).map_err(|e| DataError::ResourceError(Box::new(e)))?,
+            );
+        };
+        self.data_provider
+            .read()
+            .unwrap()
+            .as_ref()
+            .unwrap()
+            .load(req)
+            .map(Some)
+    }
+
+    pub fn try_iter(
+        &self,
+        data_key: &DataKey,
+        cldr_paths: &'b CldrPaths,
+    ) -> Result<Option<Box<dyn Iterator<Item = DataEntry>>>, DataError> {
```

Comment:
or impl `From` for those and use `?` around. I'd love to avoid unwraps in our code.

Reply:
At least I managed to get all the runtime unwraps in one file :)

All of the unwraps in this file are in cases when the mutex (RwLock) is [poisoned](https://doc.rust-lang.org/beta/std/sync/struct.PoisonError.html).  If the mutex gets poisoned, something really weird is happening, and I want that to crash the program.

I can change them to `expect` if you'd prefer that to `unwrap`.

Reply:
I mean, I can bubble up the PoisonError.  Is that better than crashing?  I was thinking that if the only way a mutex gets poisoned is if some thread gets killed, then we should probably die too, but I can change it to handle mutex errors more gracefully.

Reply:
I think either bubbling up, or having a function that returns a single failure and then `expect`-ing it will read better.

Having an infallible function with a bunch of `unwraps` feels off.